SSA/Core/MLIRSyntax/Parser.lean

```diff
@@ -0,0 +1,132 @@
+import SSA.Core.MLIRSyntax.AST
+import SSA.Core.MLIRSyntax.EDSL
+import Lean
+
+open Lean
+
+variable {ParseOutput : Type} [ToString ParseOutput]
+
+abbrev ParseError := String
+abbrev ParseFun : Type := Lean.Environment → String → EIO ParseError ParseOutput
+
+-- see https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Calling.20unification.20from.20meta.20to.20remove.20metavariables
+unsafe def elabIntoTermTactic {α : Type} (expectedType : Expr) (stx : Lean.Syntax) : Elab.Tactic.TacticM α := do
+  let expr ← Lean.Elab.Tactic.elabTerm stx none
+  let _ ← Meta.isDefEq (← Meta.inferType expr) expectedType
+  Elab.Term.synthesizeSyntheticMVarsNoPostponing
+  let expr ← instantiateMVars expr
+  Meta.evalExpr α expectedType expr
+
+namespace Lean.Elab.Tactic
+
+@[inline] private def TacticM.runCore (x : TacticM α) (ctx : Context) (s : State) : TermElabM (α × State) :=
+  x ctx |>.run s
+
+@[inline] private def TacticM.runCore' (x : TacticM α) (ctx : Context) (s : State) : TermElabM α :=
+  Prod.fst <$> x.runCore ctx s
+
+end Lean.Elab.Tactic
+
+unsafe def elabIntoTermElab {α : Type} (expectedType : Expr) (stx : Lean.Syntax) : Elab.Term.TermElabM α := do
+  elabIntoTermTactic (α := α) expectedType stx  |>.runCore' { elaborator := `ParserHack} default
+
+unsafe def elabIntoMeta {α : Type} (expectedType : Expr) (stx : Lean.Syntax) : MetaM α :=
+  elabIntoTermElab (α := α) expectedType stx |>.run'
+
+unsafe def elabIntoCore {α : Type} (expectedType : Expr) (stx : Lean.Syntax) : CoreM α :=
+  elabIntoMeta (α := α) expectedType stx |>.run'
+
+def printException : Except Exception α → IO String
+  | Except.ok _ => throw <| IO.userError "printException called on Except.ok"
+  | Except.error e => e.toMessageData.toString
+
+unsafe def elabIntoEIO {α : Type} (env : Lean.Environment) (typeName : Lean.Expr) (stx : Lean.Syntax) : EIO ParseError α :=
+  fun s =>
+    let resE : EIO Exception α := elabIntoCore (α := α) typeName stx |>.run' {fileName := "parserHack", fileMap := default} {env := env}
+    match resE s with
+    | .ok a s => .ok a s
+    | .error exception s =>
+      let errMsgIO : IO String := exception.toMessageData.toString
+      match errMsgIO s with
+      | .ok errMsg s => .error s!"failed elaboration {stx}. Error: {errMsg}" s
+      | .error e s => .error s!" Failed elaborating {stx}.\nUnable to pretty-print exception at 'elabIntoEIO':\n{e}." s
+
+def region0Expr := (Expr.app (Expr.const `MLIR.AST.Region []) (Expr.const `Nat.zero []))
+def op0Expr := (Expr.app (Expr.const `MLIR.AST.Op []) (Expr.const `Nat.zero []))
```

Comment:
Newline?

Reply:
I don't quite understand this comment. Do you want a newline between the two definitions, or a newline at the end of the file? or do you want something else?

More generally, it sounds like a sort of style question, but we don't have an agreed-upon, fix code style. I don't particularly care for this, happy to commit a suggestion (or you can push a stylistic change that you want to)

Reply:
resolved, we agree on mathlib style that uses a newline.